File: 백준/Gold/3190. 뱀/뱀.py
import sys
from collections import deque


class Snake:
    def __init__(self) -> None:
        self.pos = deque([(1, 1)])
        self.body = set(self.pos)
        self.direction = [0, 1]

    def move(self, N: int) -> bool:
        head = tuple(map(sum, zip(self.pos[0], self.direction)))
        x, y = head
        if 0 < x <= N and 0 < y <= N and head not in self.body:
            self.pos.appendleft(head)
            self.body.add(head)
            return True
        return False

    def eat(self, apple: set[tuple[int, int]]) -> None:
        head = self.pos[0]
        if head in apple:
            apple.discard(head)
        else:
            self.body.discard(self.pos.pop())

    def turn(self, d: str) -> None:
        dx, dy = self.direction
        if d == 'L':
            self.direction = [-dy, dx]
        elif d == 'D':
            self.direction = [dy, -dx]


class Dummy:
    def __init__(self) -> None:
        self.N: int
        self.K: int
        self.L: int
        self.apple: set[tuple[int, int]]
        self.snake: Snake
        self.info: dict[int, str] = {}
        self.get_info()

    def get_info(self) -> None:
        self.N = int(input())
        self.K = int(input())
        self.apple = set([tuple(map(int, input().split()))
                         for _ in range(self.K)])
        self.L = int(input())
        for _ in range(self.L):
            X, C = input().strip().split()
            self.info[int(X)] = C
        self.snake = Snake()

    def play(self) -> int:
        sec = 0
        snake = self.snake
        while True:
            sec += 1
            if snake.move(self.N) == False:
                break
            snake.eat(self.apple)
            if sec in self.info:
                snake.turn(self.info[sec])
        return sec


def main():
    game = Dummy()
    print(game.play())


if __name__ == '__main__':
    input = sys.stdin.readline
    main()

File: 백준/Gold/3190. 뱀/뱀.cc
#include <iostream>
#include <unordered_set>
#include <deque>
#include <unordered_map>

using namespace std;

bool isin(unordered_map<int, unordered_set<int>>* S, int x, int y) {
	return S->count(x) && (*S)[x].count(y) ? true : false;
}

struct hashFunction {
	size_t operator()(const pair<int, int>& x) const {
		return x.first ^ x.second;
	}

};

class Snake {
private:
	deque<pair<int, int>> pos;
	unordered_map<int, unordered_set<int>> body;
	pair<int, int> direction;
	
public:
	Snake() {
		pos.push_back({1, 1});
		body[1].insert(1);
		direction = {0, 1};
	}

	bool move(int N) {
		pair<int, int> head = pos[0];
		int x = head.first + direction.first,
			y  = head.second + direction.second;

		if (0 < x && x <= N && 0 < y && y <= N
			&& !isin(&body, x, y)) {
			pos.push_front({x, y});
			body[x].insert(y);

			return true;
		}

		return false;
	}

	void eat(unordered_map<int, unordered_set<int>>* apples) {
		pair<int, int> head = pos[0];
		int x = head.first, y = head.second;

		if (isin(apples, x, y))
			apples->at(x).erase(y);
		else {
			int tx = pos.back().first,
				ty = pos.back().second;
			body[tx].erase(ty);
			pos.pop_back();
		}
	}

	void turn(char d) {
		int dx = direction.first,
			dy = direction.second;

		if (d == 'L')
			direction.first = -dy,
			direction.second = dx;
		else if (d == 'D')
			direction.first = dy,
			direction.second = -dx;
	}
};

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL), cout.tie(NULL);

	int N, K;
	cin >> N >> K;

	unordered_map<int, unordered_set<int>> apples;
	auto* apples_ptr = &apples;
	Snake snake;
	int sec = 0;

	while (K--) {
		int x, y;
		cin >> x >> y;
		apples[x].insert(y);
	}

	int L;
	cin >> L;

	unordered_map<int, char> turn_infos;
	for (int i = 0; i < L; i++) {
		int x;
		char d;
		cin >> x >> d;
		turn_infos[x] = d;
	}

	while (true) {
		sec += 1;
		if (snake.move(N) == false) break;
		snake.eat(apples_ptr);

		if (turn_infos.count(sec))
			snake.turn(turn_infos[sec]);
	}

	cout << sec;

	return 0;
}